TomoX: RedblackTree nil exception

```go
DEBUG[08-17|04:12:49] After matching, order (unmatched part) is now back to asks tree order="&{Quantity:+2000000000000000000 Price:+974999000000000000000000000000000000000000 ExchangeAddress:[241 204 244 216 253 152 162 9 249 93 108 67 177 6 183 59 202 74 112 232] UserAddress:[191 135 91 225 96 147 167 190 51 255 150 187 105 110 241 164 11 185 117 212] BaseToken:[77 126 162 206 148 146 22 214 177 32 243 170 16 22 65 115 97 90 43 108] QuoteToken:[194 250 27 169 11 21 227 97 46 0 103 160 2 1 146 147 135 132 217 197] Status:OPEN Side:SELL Type:LO Hash:[20 171 55 138 205 209 67 24 104 108 134 159 91 35 64 176 214 69 13 82 179 12 104 135 236 244 64 128 61 78 246 57] Signature:0xc00f70eeb0 FilledAmount:+0 Nonce:+30 MakeFee:+1 TakeFee:+1 PairName:BTC/ETH CreatedAt:1566015158 UpdatedAt:1566015158 OrderID:30 NextOrder:[] PrevOrder:[] OrderList:[] Key:}"
DEBUG[08-17|04:12:49] create price list                        detail=974999000000000000000000000000000000000000
DEBUG[08-17|04:12:49] Save orderlist                           key=6274632f657468d77306f062d72d90bf7637e137203e8ddd86c838aec2c68572ce08dfc57e5708                                        value="[248 87 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 128 146 11 49 67 184 252 25 76 245 248 209 159 168 210 112 0 0 0 0]"
DEBUG[08-17|04:12:49] Key not found                            key=6274632f657468d77306f062d72d90bf7637e13720335f10ff51d711a305d26ffb83dfc57e5708                                        err="leveldb: not found"
leveldb: not found
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xb5ed2a]

goroutine 65 [running]:
github.com/ethereum/go-ethereum/tomox.(*Tree).Put(0xc003e043c0, 0xc00f45e960, 0x27, 0x30, 0xc024d4a8a0, 0x59, 0x59, 0x1, 0x421b78, 0x30)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/redblacktree.go:79 +0x70a
github.com/ethereum/go-ethereum/tomox.(*OrderTree).SaveOrderList(0xc033aeab90, 0xc0292867e0, 0xc029286701, 0x1b256d0, 0xc00732b5a0)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/ordertree.go:183 +0x474
github.com/ethereum/go-ethereum/tomox.(*OrderList).Save(...)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderlist.go:161
github.com/ethereum/go-ethereum/tomox.(*OrderTree).CreatePrice(0xc033aeab90, 0xc00a293de0, 0x1, 0x4)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/ordertree.go:167 +0x5a
github.com/ethereum/go-ethereum/tomox.(*OrderTree).InsertOrder(0xc033aeab90, 0xc0254ae160, 0x1, 0x4, 0xc002678c80)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/ordertree.go:237 +0x152
github.com/ethereum/go-ethereum/tomox.(*OrderBook).processLimitOrder(0xc033aeaaf0, 0xc0254ae160, 0x101, 0x4, 0xc002678fa8, 0x2, 0x2, 0xc02f47c801, 0xc007167168)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderbook.go:311 +0x8e4
github.com/ethereum/go-ethereum/tomox.(*OrderBook).ProcessOrder(0xc033aeaaf0, 0xc0254ae160, 0x9e5c332037e10101, 0x636f88247f65a51, 0x108577ec5cff0ea, 0x0, 0x0, 0x400, 0x35)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderbook.go:341 +0x159
github.com/ethereum/go-ethereum/tomox.(*TomoX).ProcessOrderPending(0xc0000df320, 0x1242391, 0x1e, 0x4)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/tomox.go:853 +0x7fd
github.com/ethereum/go-ethereum/miner.(*worker).commitNewWork(0xc0000b68c0)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:615 +0x1ac5
github.com/ethereum/go-ethereum/miner.(*worker).update(0xc0000b68c0)
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:287 +0x762
created by github.com/ethereum/go-ethereum/miner.newWorker
        /home/tuna/gocode/tomox/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:170 +0x405

```

--- a/tomox/orderbook.go
+++ b/tomox/orderbook.go
@@ -134,7 +134,7 @@ func (orderBook *OrderBook) Restore(dryrun bool) error {
 		log.Error("can't save asks changes to DB", "err", err)
 		return err
 	}
-	log.Debug("restored orderbook asks", "asks.Item", orderBook.Bids.Item)
+	log.Debug("restored orderbook asks", "asks.Item", orderBook.Asks.Item)
 
 	if err := orderBook.Bids.Restore(dryrun); err != nil {
 		log.Error("can't restore orderbook bids", "err", err)

--- a/tomox/redblacktree.go
+++ b/tomox/redblacktree.go
@@ -186,6 +186,19 @@ func (tree *Tree) Remove(key []byte, dryrun bool) {
 			child = right
 		}
 		if child == nil {
+			parent := node.Parent(tree, dryrun)
+			if parent == nil {
+				log.Error("Parent can't be nil", "node.Key", hex.EncodeToString(node.Key))
+				return
+			}
+			if !tree.IsEmptyKey(parent.LeftKey()) && bytes.Equal(parent.LeftKey(), node.Key) {
+				parent.Item.Keys.Left = EmptyKey()
+			} else if !tree.IsEmptyKey(parent.RightKey()) && bytes.Equal(parent.RightKey(), node.Key) {
+				parent.Item.Keys.Right = EmptyKey()
+			} else {
+				log.Error("Parent doesn't have node as child", "node.Key", hex.EncodeToString(node.Key))
+				return
+			}
 			tree.deleteNode(node, dryrun)
 			log.Debug("Removed node with child = nil", "node", hex.EncodeToString(node.Key), "all keys", tree.KeysinString(true))
 		} else {

--- a/tomox/redblacktree.go
+++ b/tomox/redblacktree.go
@@ -187,21 +187,19 @@ func (tree *Tree) Remove(key []byte, dryrun bool) {
 		}
 		if child == nil {
 			parent := node.Parent(tree, dryrun)
-			if parent == nil {
-				log.Error("Parent can't be nil", "node.Key", hex.EncodeToString(node.Key))
-				return
-			}
-			if !tree.IsEmptyKey(parent.LeftKey()) && bytes.Equal(parent.LeftKey(), node.Key) {
-				parent.Item.Keys.Left = EmptyKey()
-				log.Debug("Set parent.LeftKey() to nil", "parent.Key", parent.Item.Keys)
-				tree.Save(parent, dryrun)
-			} else if !tree.IsEmptyKey(parent.RightKey()) && bytes.Equal(parent.RightKey(), node.Key) {
-				parent.Item.Keys.Right = EmptyKey()
-				log.Debug("Set parent.RightKey() to nil", "parent.Key", parent.Item.Keys)
-				tree.Save(parent, dryrun)
-			} else {
-				log.Error("Parent doesn't have node as child", "node.Key", hex.EncodeToString(node.Key))
-				return
+			if parent != nil {
+				if !tree.IsEmptyKey(parent.LeftKey()) && bytes.Equal(parent.LeftKey(), node.Key) {
+					parent.Item.Keys.Left = EmptyKey()
+					log.Debug("Set parent.LeftKey() to nil", "parent.Key", parent.Item.Keys)
+					tree.Save(parent, dryrun)
+				} else if !tree.IsEmptyKey(parent.RightKey()) && bytes.Equal(parent.RightKey(), node.Key) {
+					parent.Item.Keys.Right = EmptyKey()
+					log.Debug("Set parent.RightKey() to nil", "parent.Key", parent.Item.Keys)
+					tree.Save(parent, dryrun)
+				} else {
+					log.Error("Parent doesn't have node as child", "node.Key", hex.EncodeToString(node.Key))
+					return
+				}
 			}
 			tree.deleteNode(node, dryrun)
 			log.Debug("Removed node with child = nil", "node", hex.EncodeToString(node.Key), "all keys", tree.KeysinString(true))

--- a/tomox/redblacktree.go
+++ b/tomox/redblacktree.go
@@ -193,8 +193,12 @@ func (tree *Tree) Remove(key []byte, dryrun bool) {
 			}
 			if !tree.IsEmptyKey(parent.LeftKey()) && bytes.Equal(parent.LeftKey(), node.Key) {
 				parent.Item.Keys.Left = EmptyKey()
+				log.Debug("Set parent.LeftKey() to nil", "parent.Key", parent.Item.Keys)
+				tree.Save(parent, dryrun)
 			} else if !tree.IsEmptyKey(parent.RightKey()) && bytes.Equal(parent.RightKey(), node.Key) {
 				parent.Item.Keys.Right = EmptyKey()
+				log.Debug("Set parent.RightKey() to nil", "parent.Key", parent.Item.Keys)
+				tree.Save(parent, dryrun)
 			} else {
 				log.Error("Parent doesn't have node as child", "node.Key", hex.EncodeToString(node.Key))
 				return

